File: src/gg/obsidian/mobcontrol/MobControl.java
package gg.obsidian.mobcontrol;

import org.bukkit.command.Command;
import org.bukkit.command.CommandSender;
import org.bukkit.entity.*;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.Listener;
import org.bukkit.event.entity.CreatureSpawnEvent;
import org.bukkit.event.world.ChunkUnloadEvent;
import org.bukkit.plugin.java.JavaPlugin;
import org.bukkit.util.Vector;

import java.io.File;

public class MobControl extends JavaPlugin implements Listener {

	private MobControlListener listener = new MobControlListener(this);
	public final Configuration config = new Configuration(this);
	public final MobManager mobs = new MobManager(this);

	public void onEnable() {
		File configFile = new File(getDataFolder(), "config.yml");
		if (!configFile.exists()) {
			getConfig().options().copyDefaults(true);
			saveConfig();
		}

		config.load();

		getServer().getPluginManager().registerEvents(listener, this);
	}

	public void onDisable() {
		mobs.removeFromAllChunks();
	}

	public boolean onCommand(CommandSender sender, Command cmd, String label, String[] args) {
		if (!cmd.getName().equalsIgnoreCase("mobcontrol")) {
			return true;
		}

		if (args.length == 1 && args[0].equalsIgnoreCase("cull") && sender.hasPermission("mobcontrol.cull")) {
			sender.sendMessage("[MobControl] Culling mobs in all loaded chunks");
			mobs.removeFromAllChunks();
			return true;
		}

		if (args.length == 1 && args[0].equalsIgnoreCase("check") && sender.hasPermission("mobcontrol.check") && sender instanceof Player) {
			LivingEntity living = (LivingEntity) getTarget((Player) sender);
			if (living == null) {
				sender.sendMessage("[MobLimiter] No mob in sight");
				return true;
			}

			boolean animalOrMonster = (living instanceof Animals) || (living instanceof Monster);
			if (!mobs.isSpecialMob(living) && animalOrMonster) {
				sender.sendMessage("[MobLimiter] Mob will be culled");
			} else {
				sender.sendMessage("[MobLimiter] Mob will not be culled");
			}

			return true;
		}

		return true;
	}

	private static Entity getTarget(final Player player) {
		assert player != null;
		Entity target = null;
		double targetDistanceSquared = 0;
		final double radiusSquared = 1;
		final Vector l = player.getEyeLocation().toVector();
		final Vector n = player.getLocation().getDirection().normalize();
		final double cos45 = Math.cos(Math.PI / 4);

		for (final LivingEntity other : player.getWorld().getEntitiesByClass(LivingEntity.class)) {
			if (other == player) {
				continue;
			}

			if (target == null | targetDistanceSquared > other.getLocation().distanceSquared(player.getLocation())) {
				final Vector t = other.getLocation().add(0, 1, 0).toVector().subtract(l);
				if (n.clone().crossProduct(t).lengthSquared() < radiusSquared && t.normalize().dot(n) >= cos45) {
					target = other;
					targetDistanceSquared = target.getLocation().distanceSquared(player.getLocation());
				}
			}
		}

		return target;
	}
}

File: src/main/kotlin/gg/obsidian/mobcontrol/MobControl.kt
package gg.obsidian.mobcontrol

import org.bukkit.command.Command
import org.bukkit.command.CommandSender
import org.bukkit.entity.*
import org.bukkit.plugin.java.JavaPlugin

import java.io.File

public class MobControl() : JavaPlugin() {

    private val listener = MobControlListener(this)
    val config = Configuration(this)
    val mobs = MobManager(this)

    override fun onEnable() {
        val configFile = File(dataFolder, "config.yml")
        if (!configFile.exists()) {
            getConfig().options().copyDefaults(true)
            saveConfig()
        }

        config.load()

        server.pluginManager.registerEvents(listener, this)
    }

    override fun onDisable() {
        mobs.removeFromAllChunks()
    }

    override fun onCommand(sender: CommandSender?, cmd: Command?, label: String?, args: Array<String>?): Boolean {
        if (!cmd!!.name.equals("mobcontrol", ignoreCase = true)) {
            return true
        }

        if (args!!.size() == 1 && args[0].equals("cull", ignoreCase = true) && sender!!.hasPermission("mobcontrol.cull")) {
            sender!!.sendMessage("[MobControl] Culling mobs in all loaded chunks")
            mobs.removeFromAllChunks()
            return true
        }

        if (args.size == 1 && args[0].equals("check", ignoreCase = true) && sender!!.hasPermission("mobcontrol.check") && sender is Player) {
            val living = getTarget(sender as Player?) as LivingEntity
            if (living == null) {
                sender.sendMessage("[MobLimiter] No mob in sight")
                return true
            }

            val animalOrMonster = (living is Animals) || (living is Monster)
            if (!mobs.isSpecialMob(living) && animalOrMonster) {
                sender.sendMessage("[MobLimiter] Mob will be culled")
            } else {
                sender.sendMessage("[MobLimiter] Mob will not be culled")
            }

            return true
        }

        return true
    }

    private fun getTarget(player: Player?): Entity? {
        assert(player != null)
        var target: Entity? = null
        var targetDistanceSquared = 0.0
        val radiusSquared = 1.0
        val l = player!!.eyeLocation.toVector()
        val n = player.location.direction.normalize()
        val cos45 = Math.cos(Math.PI / 4)

        for (other in player.world.getEntitiesByClass(LivingEntity::class.java)) {
            if (other === player) {
                continue
            }

            if ((target == null) or (targetDistanceSquared > other.location.distanceSquared(player.location))) {
                val t = other.location.add(0.0, 1.0, 0.0).toVector().subtract(l)
                if (n.clone().crossProduct(t).lengthSquared() < radiusSquared && t.normalize().dot(n) >= cos45) {
                    target = other
                    targetDistanceSquared = target!!.location.distanceSquared(player.location)
                }
            }
        }

        return target
    }
}
